draft-ietf-tls-tls13.md

```diff
@@ -1984,11 +1985,25 @@ KeyShareEntry values for groups not listed in the client's
 these rules and and MAY abort the connection with a fatal
 "illegal_parameter" alert if one is violated.
 
+Upon receipt of this extension in a HelloRetryRequest, the client MUST first
+verify that the selected_group field corresponds to a group which was provided
+in the "supported_groups" extension in the original ClientHello. It MUST then
+verify that the selected_group field does not correspond to a group which was
+provided in the "key_share" extension in the original ClientHello. If either of
+these checks fails, then the client MUST abort the handshake with a fatal
+"illegal_parameter" alert.  Otherwise, when sending the new ClientHello, the
+client MUST append a new KeyShareEntry for the group indicated in the
+selected_group field to the groups in its original KeyShare. The remaining
+KeyShareEntry values MUST be preserved.
+
 If using (EC)DHE key establishment, servers offer exactly one
-KeyShareEntry. This value MUST correspond to the KeyShareEntry value offered
+KeyShareEntry in the ServerHello. This value MUST correspond to the KeyShareEntry value offered
 by the client that the server has selected for the negotiated key exchange.
 Servers MUST NOT send a KeyShareEntry for any group not
 indicated in the "supported_groups" extension.
+If a HelloRetryRequest was received, the client MUST verify that the
+selected NamedGroup matches that supplied in the selected_group field and MUST
```

Comment:
Maybe:

> If a HelloRetryRequest included the supported_groups extension, the client MUST verify that the key_share value in the ServerHello is for the same group and abort [...]

Note that HRR might not include supported_groups if the server is only sending a cookie.